ccl/changefeedccl: TestChangefeedOnlyInitialScan failed

ccl/changefeedccl.TestChangefeedOnlyInitialScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9497764?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9497764?buildTab=artifacts#/) on release-23.1 @ [de239a7438f44d382c9aefceb65d9c39911dabd2](https://github.com/cockroachdb/cockroach/commits/de239a7438f44d382c9aefceb65d9c39911dabd2):


```
=== RUN   TestChangefeedOnlyInitialScan
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/d2b64858287e5594aed9da7514006ac9/logTestChangefeedOnlyInitialScan818653931
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestChangefeedOnlyInitialScan
    changefeed_test.go:7141: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d2b64858287e5594aed9da7514006ac9/logTestChangefeedOnlyInitialScan818653931
--- FAIL: TestChangefeedOnlyInitialScan (1.93s)
=== RUN   TestChangefeedOnlyInitialScan/pubsub
    helpers_test.go:820: making server as system tenant
    helpers_test.go:895: making pubsub feed factory
    --- FAIL: TestChangefeedOnlyInitialScan/pubsub (1.92s)
=== RUN   TestChangefeedOnlyInitialScan/pubsub/initial_backfill_only
    changefeed_test.go:7110: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
        --- FAIL: TestChangefeedOnlyInitialScan/pubsub/initial_backfill_only (0.26s)
=== RUN   TestChangefeedOnlyInitialScan/pubsub/initial_scan_only
    changefeed_test.go:7110: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
        --- FAIL: TestChangefeedOnlyInitialScan/pubsub/initial_scan_only (0.26s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #94980 ccl/changefeedccl: TestChangefeedOnlyInitialScan failed [C-test-failure O-robot T-cdc branch-release-22.2]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedOnlyInitialScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26714

ccl/changefeedccl.TestChangefeedOnlyInitialScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9525940?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9525940?buildTab=artifacts#/) on release-23.1 @ [f4145091448687af29b54054a61c32f96303af96](https://github.com/cockroachdb/cockroach/commits/f4145091448687af29b54054a61c32f96303af96):


```
=== RUN   TestChangefeedOnlyInitialScan
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/d2b64858287e5594aed9da7514006ac9/logTestChangefeedOnlyInitialScan4178281339
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestChangefeedOnlyInitialScan
    changefeed_test.go:7141: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d2b64858287e5594aed9da7514006ac9/logTestChangefeedOnlyInitialScan4178281339
I230410 15:02:19.000459 69777 3@pebble/event.go:706  [n?,s?,pebble] 361  [JOB 1] MANIFEST created 000001
I230410 15:02:19.000672 69777 3@pebble/event.go:734  [n?,s?,pebble] 362  [JOB 1] WAL created 000002
I230410 15:02:19.000724 69777 3@pebble/event.go:702  [n?,s?,pebble] 363  upgraded to format version: 002
I230410 15:02:19.000752 69777 3@pebble/event.go:702  [n?,s?,pebble] 364  upgraded to format version: 003
I230410 15:02:19.000769 69777 3@pebble/event.go:702  [n?,s?,pebble] 365  upgraded to format version: 004
I230410 15:02:19.000783 69777 3@pebble/event.go:702  [n?,s?,pebble] 366  upgraded to format version: 005
I230410 15:02:19.000800 69777 3@pebble/event.go:702  [n?,s?,pebble] 367  upgraded to format version: 006
I230410 15:02:19.000816 69777 3@pebble/event.go:702  [n?,s?,pebble] 368  upgraded to format version: 007
I230410 15:02:19.000833 69777 3@pebble/event.go:702  [n?,s?,pebble] 369  upgraded to format version: 008
I230410 15:02:19.000849 69777 3@pebble/event.go:702  [n?,s?,pebble] 370  upgraded to format version: 009
I230410 15:02:19.000865 69777 3@pebble/event.go:702  [n?,s?,pebble] 371  upgraded to format version: 010
I230410 15:02:19.000976 69777 server/config.go:847  [T1,n?] 372  1 storage engine initialized
I230410 15:02:19.000996 69777 server/config.go:850  [T1,n?] 373  Pebble cache size: 128 MiB
I230410 15:02:19.001010 69777 server/config.go:850  [T1,n?] 374  store 0: in-memory, size 512 MiB
I230410 15:02:19.001022 69777 server/config.go:850  [T1,n?] 375  store 0: {Encrypted:false ReadOnly:false FileStoreProperties:<nil>}
--- FAIL: TestChangefeedOnlyInitialScan (2.73s)
=== RUN   TestChangefeedOnlyInitialScan/pubsub
    helpers_test.go:820: making server as system tenant
    helpers_test.go:895: making pubsub feed factory
    --- FAIL: TestChangefeedOnlyInitialScan/pubsub (2.70s)
=== RUN   TestChangefeedOnlyInitialScan/pubsub/initial_backfill_only
    changefeed_test.go:7110: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
        --- FAIL: TestChangefeedOnlyInitialScan/pubsub/initial_backfill_only (0.33s)
=== RUN   TestChangefeedOnlyInitialScan/pubsub/initial_scan_only
    changefeed_test.go:7110: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
        --- FAIL: TestChangefeedOnlyInitialScan/pubsub/initial_scan_only (0.33s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #94980 ccl/changefeedccl: TestChangefeedOnlyInitialScan failed [C-test-failure O-robot T-cdc branch-release-22.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedOnlyInitialScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/sink_pubsub_v2.go
+++ b/pkg/ccl/changefeedccl/sink_pubsub_v2.go
@@ -104,9 +104,16 @@ func makePubsubSinkClient(
 		return nil, errors.New("missing project name")
 	}
 
-	publisherClient, err := makePublisherClient(ctx, pubsubURL, unordered, knobs)
-	if err != nil {
-		return nil, err
+	var err error
+	var publisherClient *pubsub.PublisherClient
+
+	// In unit tests the publisherClient gets set immediately after initializing
+	// the sink object via knobs.WrapSink.
+	if knobs == nil || !knobs.PubsubClientSkipClientCreation {
+		publisherClient, err = makePublisherClient(ctx, pubsubURL, unordered)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	sinkClient := &pubsubSinkClient{
@@ -250,7 +257,7 @@ func (pe *pubsubSinkClient) Close() error {
 }
 
 func makePublisherClient(
-	ctx context.Context, url sinkURL, unordered bool, knobs *TestingKnobs,
+	ctx context.Context, url sinkURL, unordered bool,
 ) (*pubsub.PublisherClient, error) {
 	const regionParam = "region"
 	region := url.consumeParam(regionParam)
@@ -267,21 +274,15 @@ func makePublisherClient(
 		endpoint = gcpEndpointForRegion(region)
 	}
 
-	options := []option.ClientOption{
-		option.WithEndpoint(endpoint),
-	}
-
-	if knobs == nil || !knobs.PubsubClientSkipCredentialsCheck {
-		creds, err := getGCPCredentials(ctx, url)
-		if err != nil {
-			return nil, err
-		}
-		options = append(options, creds)
+	creds, err := getGCPCredentials(ctx, url)
+	if err != nil {
+		return nil, err
 	}
 
 	client, err := pubsub.NewPublisherClient(
 		ctx,
-		options...,
+		option.WithEndpoint(endpoint),
+		creds,
 	)
 	if err != nil {
 		return nil, errors.Wrap(err, "opening client")

--- a/pkg/ccl/changefeedccl/testfeed_test.go
+++ b/pkg/ccl/changefeedccl/testfeed_test.go
@@ -2359,11 +2359,11 @@ func makePubsubFeedFactory(srvOrCluster interface{}, db *gosql.DB) cdctest.TestF
 
 	switch t := srvOrCluster.(type) {
 	case serverutils.TestTenantInterface:
-		t.DistSQLServer().(*distsql.ServerImpl).TestingKnobs.Changefeed.(*TestingKnobs).PubsubClientSkipCredentialsCheck = true
+		t.DistSQLServer().(*distsql.ServerImpl).TestingKnobs.Changefeed.(*TestingKnobs).PubsubClientSkipClientCreation = true
 	case serverutils.TestClusterInterface:
 		servers := make([]feedInjectable, t.NumServers())
 		for i := range servers {
-			t.Server(i).DistSQLServer().(*distsql.ServerImpl).TestingKnobs.Changefeed.(*TestingKnobs).PubsubClientSkipCredentialsCheck = true
+			t.Server(i).DistSQLServer().(*distsql.ServerImpl).TestingKnobs.Changefeed.(*TestingKnobs).PubsubClientSkipClientCreation = true
 		}
 	}
 
@@ -2404,8 +2404,6 @@ func (p *pubsubFeedFactory) Feed(create string, args ...interface{}) (cdctest.Te
 		defer mu.Unlock()
 		if batchingSink, ok := s.(*batchingSink); ok {
 			if sinkClient, ok := batchingSink.client.(*pubsubSinkClient); ok {
-				_ = sinkClient.client.Close()
-
 				conn, _ := mockServer.Dial()
 				mockClient, _ := pubsubv1.NewPublisherClient(context.Background(), option.WithGRPCConn(conn))
 				sinkClient.client = mockClient

--- a/pkg/ccl/changefeedccl/testing_knobs.go
+++ b/pkg/ccl/changefeedccl/testing_knobs.go
@@ -32,8 +32,9 @@ type TestingKnobs struct {
 	// It allows the tests to muck with the Sink, and even return altogether different
 	// implementation.
 	WrapSink func(s Sink, jobID jobspb.JobID) Sink
-	// PubsubClientSkipCredentialsCheck, if set, skips the gcp credentials checking
-	PubsubClientSkipCredentialsCheck bool
+	// PubsubClientSkipClientCreation, if set, skips creating a google cloud
+	// client as it is expected that the test manually sets a client.
+	PubsubClientSkipClientCreation bool
 	// FilterSpanWithMutation is a filter returning true if the resolved span event should
 	// be skipped. This method takes a pointer in case resolved spans need to be mutated.
 	FilterSpanWithMutation func(resolved *jobspb.ResolvedSpan) bool

